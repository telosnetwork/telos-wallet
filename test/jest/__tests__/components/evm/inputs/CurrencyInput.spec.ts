import CurrencyInput from 'components/evm/inputs/CurrencyInput.vue';
import { shallowMount } from '@vue/test-utils';
import { BigNumber } from 'ethers';
import { oneEthInWei, oneThousandFiveHundredEthInWei, stubWithSlot } from 'test/jest/testing-helpers';

describe('CurrencyInput.vue', () => {
    const stubs = {
        'ToolTip': stubWithSlot('ToolTip'),
        'q-spinner': stubWithSlot('q-spinner'),
    };

    it('should have the correct name', () => {
        expect(CurrencyInput.name).toEqual('CurrencyInput');
    });

    it('should emit the correct values when currencies are not swapped', async () => {
        const wrapper = shallowMount(CurrencyInput, {
            global: {
                stubs,
            },
            propsData: {
                modelValue: BigNumber.from(0),
                symbol: 'TLOS',
                decimals: 18,
                secondaryCurrencyConversionFactor: '0.19',
                secondaryCurrencySymbol: 'USD',
                secondaryCurrencyDecimals: 2,
                locale: 'en-US',
                label: 'TLOS / USD',
            },
        });
        const inputElement = wrapper.find('input');

        expect(wrapper.element).toMatchSnapshot(); // blank
        await inputElement.setValue('1');
        expect(wrapper.emitted('update:modelValue')?.length).toBe(1);
        expect(wrapper.emitted('update:modelValue')?.[0]).toEqual([BigNumber.from(oneEthInWei)]);

        await wrapper.setProps({
            modelValue: BigNumber.from(oneEthInWei),
        });

        // no new value should be emitted when modelValue is updated to the same value as was emitted
        expect(wrapper.emitted('update:modelValue')?.length).toBe(1);
        expect(wrapper.element).toMatchSnapshot(); // 1 in the input, 0.19 in the secondary currency
        expect(inputElement.element.value).toBe('1');

        await inputElement.setValue('1,500');
        expect(wrapper.emitted('update:modelValue')?.length).toBe(2);
        expect(wrapper.emitted('update:modelValue')?.[1]).toEqual([BigNumber.from(oneThousandFiveHundredEthInWei)]);

        await wrapper.setProps({
            modelValue: BigNumber.from(oneThousandFiveHundredEthInWei),
        });

        expect(wrapper.emitted('update:modelValue')?.length).toBe(2);
        expect(wrapper.element).toMatchSnapshot(); // 1,500 in the input, 285 in the secondary currency
        expect(inputElement.element.value).toBe('1,500');
    });

    it('should emit the correct values when currencies are swapped', async () => {
        const wrapper = shallowMount(CurrencyInput, {
            global: {
                stubs,
            },
            propsData: {
                modelValue: BigNumber.from(0),
                symbol: 'TLOS',
                decimals: 18,
                secondaryCurrencyConversionFactor: '0.2',
                secondaryCurrencySymbol: 'USD',
                secondaryCurrencyDecimals: 2,
                locale: 'en-US',
                label: 'TLOS / USD',
            },
        });
        const inputElement = wrapper.find('input');

        await wrapper.find('.c-currency-input__currency-switcher').trigger('click');
        expect(wrapper.emitted('update:modelValue')?.length).toBe(undefined);

        await inputElement.setValue('0.2');

        expect(wrapper.emitted('update:modelValue')?.length).toBe(1);
        expect(wrapper.emitted('update:modelValue')?.[0]).toEqual([BigNumber.from(oneEthInWei)]);

        await wrapper.setProps({
            modelValue: BigNumber.from(oneEthInWei),
        });
        expect(wrapper.emitted('update:modelValue')?.length).toBe(1);
        expect(wrapper.element).toMatchSnapshot(); // 0.2 in the input, 1 in the secondary currency
    });
});
