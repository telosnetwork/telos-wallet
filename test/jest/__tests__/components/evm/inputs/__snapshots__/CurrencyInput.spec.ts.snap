// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyInput.vue should emit the correct values when currencies are not swapped 1`] = `
<div
  class="c-currency-input c-text-input c-text-input--error-right"
>
  <div
    class="c-text-input__label-text"
    id="currency-input-label--test"
  >
    TLOS / USD (TLOS)
  </div>
  <!--v-if-->
  <input
    aria-labelledby="currency-input-label--test"
    class="c-text-input__input"
    inputmode="decimal"
    pattern="[0-9,.]*"
    placeholder="0"
    type="text"
  />
  <q-spinner-stub
    class="c-currency-input__spinner"
    color="primary"
    size="md"
    style="display: none;"
  >
    
    
  </q-spinner-stub>
  <div
    class="c-currency-input__currency-switcher"
  >
    <inline-svg-stub
      aria-hidden="true"
      class="c-currency-input__swap-icon"
      keepduringloading="true"
      src=""
      transformsource="[Function]"
    />
     0.00 USD
  </div>
  <div
    class="c-text-input__error-text"
  />
  <div
    class="c-currency-input__conversion-rate"
  >
    @ 0.19 USD / TLOS
  </div>
</div>
`;

exports[`CurrencyInput.vue should emit the correct values when currencies are not swapped 2`] = `
<div
  class="c-currency-input c-text-input c-text-input--error-right"
>
  <div
    class="c-text-input__label-text"
    id="currency-input-label--test"
  >
    TLOS / USD (TLOS)
  </div>
  <!--v-if-->
  <input
    aria-labelledby="currency-input-label--test"
    class="c-text-input__input"
    inputmode="decimal"
    pattern="[0-9,.]*"
    placeholder="0"
    type="text"
  />
  <q-spinner-stub
    class="c-currency-input__spinner"
    color="primary"
    size="md"
    style="display: none;"
  >
    
    
  </q-spinner-stub>
  <div
    class="c-currency-input__currency-switcher"
  >
    <inline-svg-stub
      aria-hidden="true"
      class="c-currency-input__swap-icon"
      keepduringloading="true"
      src=""
      transformsource="[Function]"
    />
     0.19 USD
  </div>
  <div
    class="c-text-input__error-text"
  />
  <div
    class="c-currency-input__conversion-rate"
  >
    @ 0.19 USD / TLOS
  </div>
</div>
`;

exports[`CurrencyInput.vue should emit the correct values when currencies are not swapped 3`] = `
<div
  class="c-currency-input c-text-input c-text-input--error-right"
>
  <div
    class="c-text-input__label-text"
    id="currency-input-label--test"
  >
    TLOS / USD (TLOS)
  </div>
  <!--v-if-->
  <input
    aria-labelledby="currency-input-label--test"
    class="c-text-input__input"
    inputmode="decimal"
    pattern="[0-9,.]*"
    placeholder="0"
    type="text"
  />
  <q-spinner-stub
    class="c-currency-input__spinner"
    color="primary"
    size="md"
    style="display: none;"
  >
    
    
  </q-spinner-stub>
  <div
    class="c-currency-input__currency-switcher"
  >
    <inline-svg-stub
      aria-hidden="true"
      class="c-currency-input__swap-icon"
      keepduringloading="true"
      src=""
      transformsource="[Function]"
    />
     285.00 USD
  </div>
  <div
    class="c-text-input__error-text"
  />
  <div
    class="c-currency-input__conversion-rate"
  >
    @ 0.19 USD / TLOS
  </div>
</div>
`;

exports[`CurrencyInput.vue should emit the correct values when currencies are swapped 1`] = `
<div
  class="c-currency-input c-text-input c-text-input--error-right"
>
  <div
    class="c-text-input__label-text"
    id="currency-input-label--test-2"
  >
    TLOS / USD (USD)
  </div>
  <!--v-if-->
  <input
    aria-labelledby="currency-input-label--test-2"
    class="c-text-input__input"
    inputmode="decimal"
    pattern="[0-9,.]*"
    placeholder="0"
    type="text"
  />
  <q-spinner-stub
    class="c-currency-input__spinner"
    color="primary"
    size="md"
    style="display: none;"
  >
    
    
  </q-spinner-stub>
  <div
    class="c-currency-input__currency-switcher"
  >
    <inline-svg-stub
      aria-hidden="true"
      class="c-currency-input__swap-icon"
      keepduringloading="true"
      src=""
      transformsource="[Function]"
    />
     1.0000 TLOS
  </div>
  <div
    class="c-text-input__error-text"
  />
  <div
    class="c-currency-input__conversion-rate"
  >
    @ 0.2 USD / TLOS
  </div>
</div>
`;
